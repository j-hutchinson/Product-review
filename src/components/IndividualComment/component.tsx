import React from 'react';
import styled from 'styled-components';
import { spacing } from '../../css/tokens';
import { Comment } from '../../types';

interface Props {
    comment: Comment;
}

const StyledComment = styled.li`
    margin: 0 ${spacing.medium};
    padding: ${spacing.medium} 0;
`;

const StyledLabel = styled.div`
    padding: 2px 0;
`;

const StyledTopLabel = styled.div`
    display: flex;
    padding: 2px 0;
`;

const StyledEnd = styled.div`
    margin-left: auto;
`;

const IndividualComment = ({ comment: { name, rating, email, comment, timePosted } }: Props): JSX.Element => (
    <StyledComment>
        <StyledTopLabel>
            <div><strong>{`Name: `}</strong>{name}</div>
            <StyledEnd><strong>{`Date: `}</strong>{timePosted}</StyledEnd>
        </StyledTopLabel>
        <StyledLabel><strong>{`Rating: `}</strong>{rating}</StyledLabel>
        <StyledLabel><strong>{`Email: `}</strong>{email}</StyledLabel>
        <StyledLabel><strong>{`Comment: `}</strong>{comment}</StyledLabel>
    </StyledComment>
);

export default IndividualComment;