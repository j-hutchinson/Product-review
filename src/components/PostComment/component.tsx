import React, { ChangeEvent, FormEvent, useState } from 'react'
import Rater from 'react-rater';
import 'react-rater/lib/react-rater.css';
import styled from 'styled-components'
import { addComment } from '../../store/actions/actions';
import store from '../../store/store';

export const StyledContainer = styled.form``;

const StyledItem = styled.div`
    display: grid;
    grid-template-rows: min-content 1fr;
    grid-gap: 4px;
    margin: 10px;
`;

export const StyledInput = styled.input`
   height: 24px;
   font-size: 14px;
`;

const StyledButtonWrapper = styled.div`
    display: flex;
    flex-direction: row-reverse;
    margin-right: 10px;
`;

const StyledSubmitInput = styled.input`
    background-color: #03fc28;
    border-radius: 4px;
    border: 1px solid black;
    font-size: 14px; 
    padding: 8px;
`;

const StyledRatingLabel = styled.p`
    padding-left: 20px;
    display: inline-block;
`;

const PostComment = (): JSX.Element => {
    const [name, setName] = useState<string>('');
    const [email, setEmail] = useState<string>('');
    const [rating, setRating] = useState<number>(0);
    const [comment, setComment] = useState<string>('');

    const onSubmission = (ev: FormEvent): void => {
        ev.preventDefault();
        store.dispatch(addComment({ name, email, rating, comment }));
        setName('');
        setEmail('');
        setRating(0);
        setComment('');
    }

    return (
        <StyledContainer onSubmit={onSubmission}>
            <StyledItem>
                <label htmlFor="name"> Name: </label>
                <StyledInput
                    id="name"
                    type="text"
                    value={name}
                    onChange={(ev: ChangeEvent<HTMLInputElement>): void => setName(ev.target.value)}
                    required
                />
            </StyledItem>
            <StyledItem>
                <label htmlFor="email"> Email: </label>
                <StyledInput
                    id="email"
                    type="email"
                    value={email}
                    onChange={(ev: ChangeEvent<HTMLInputElement>): void => setEmail(ev.target.value)}
                    required
                />
            </StyledItem>
            <StyledItem>Rating:
                <div>
                    <Rater
                        rating={rating}
                        total={5}
                        onRate={({ rating }): void => setRating(rating)}
                    />
                    <StyledRatingLabel>({rating}) stars</StyledRatingLabel>
                </div>
            </StyledItem>
            <StyledItem>
                <label htmlFor="comment"> Comment </label>
                <StyledInput
                    id="comment"
                    type="text"
                    value={comment}
                    onChange={(ev: ChangeEvent<HTMLInputElement>): void => setComment(ev.target.value)}
                    required
                />
            </StyledItem>
            <StyledButtonWrapper>
                <StyledSubmitInput type="submit" value="Submit" />
            </StyledButtonWrapper>
        </StyledContainer>
    );
}
export default PostComment