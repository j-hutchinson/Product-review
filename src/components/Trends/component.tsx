import React from 'react';
import { connect } from 'react-redux';
import { Doughnut } from 'react-chartjs-2';
import styled from 'styled-components'
import { ReduxState } from '../../types';

interface Props {
    ratings: number[]
}

const TrendContainer = styled.div`
    padding: 20px;
    border: 1px solid black;
    border-radius: 4px;
`;

export const StyledEmptyLabel = styled.div`
    text-align: center;
    margin: auto;
    padding-top: 40px;
`;

export const Trends = ({ ratings }) => {
    let one = 0; let two = 0; let three = 0; let four = 0; let five = 0;
    ratings.forEach(rating => {
        if (rating === 1) one++;
        if (rating === 2) two++;
        if (rating === 3) three++;
        if (rating === 4) four++;
        if (rating === 5) five++;
    })

    const data = {
        labels: ['1', '2', '3', '4', '5'],
        datasets: [
            {
                label: 'Ratings',
                backgroundColor: ['#FF6384', '#36A2EB', '#FFCE56', '#03FC28'],
                borderColor: 'black',
                borderWidth: 1,
                hoverBackgroundColor: 'rgba(255,99,132,0.4)',
                hoverBorderColor: 'rgba(255,99,132,1)',
                data: [one, two, three, four, five]
            }
        ]
    };

    return (
        <TrendContainer>
            {ratings.length ?
                <Doughnut
                    data={data}
                    width={100}
                    height={50}
                    options={{ maintainAspectRatio: false }}
                /> :
                <StyledEmptyLabel>No data available, try leaving a comment!</StyledEmptyLabel>
            }
        </TrendContainer>
    )
}

export const mapStateToProps = (state: ReduxState): Props => ({
    ratings: state.allComments.map(comment => comment.rating)
})

export default connect(mapStateToProps, null)(Trends);