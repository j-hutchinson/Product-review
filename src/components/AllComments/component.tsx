import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { spacing } from '../../css/tokens';
import { Comment, ReduxState } from '../../types';
import IndividualComment from '../IndividualComment/component';

interface Props {
    allComments: Comment[]
}

const StyledCommentsContainer = styled.div`
    border: 1px solid black;
    border-radius: 4px;
    padding: ${spacing.medium};
`;

const StyledStrong = styled.strong`
    padding-left: ${spacing.medium};
    font-size: 18px;
`;

const StyledComments = styled.ul`
    list-style: none;
    margin: 0;
    padding: 0;

    li:not(:last-child) {
        border-bottom: 1px solid #c7c7c7;
    }
`;

export const StyledEmptyLabel = styled.p`
    padding-left: ${spacing.medium};
`;

export const AllComments = ({ allComments }: Props): JSX.Element => (
    <StyledCommentsContainer>
        <StyledStrong>Comments:</StyledStrong>
        {allComments.length ?
            <StyledComments>
                {allComments.map((comment: Comment, ind: number): JSX.Element =>
                    <IndividualComment comment={comment} key={`${comment.name}${ind}`} />)}
            </StyledComments>
            :
            <StyledEmptyLabel>No comments available, be the first to leave a comment!</StyledEmptyLabel>
        }
    </StyledCommentsContainer>
);

export const mapStateToProps = (state: ReduxState): Props => ({
    allComments: state.allComments
})

export default connect(mapStateToProps, null)(AllComments);