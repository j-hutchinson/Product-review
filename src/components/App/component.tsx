import React from 'react';
import styled from 'styled-components';
import store from '../../store/store';
import { addMultipleComments } from '../../store/actions/actions';
import AllComments from '../AllComments/component';
import PostComment from '../PostComment/component';
import Trends from '../Trends/component';
import { mockData } from '../../data/mockData';

const StyledWrapper = styled.div`
  display: grid;
  grid-template-rows: 1fr 1fr;
  height: 275px;
  grid-gap: 40px;
  font-size: 14px;
  margin: auto;
  padding: 60px;
`;

const StyledSection = styled.div` 
  display: grid;
  grid-template-columns: 1fr 1fr;
`;

const Button = styled.button`
  width: 100px;
`;

const App = (): JSX.Element => {
  const addMockData = () => store.dispatch(addMultipleComments(mockData));

  return (
    <StyledWrapper>
      <StyledSection>
        <PostComment />
        <Trends />
      </StyledSection>
      <AllComments />
      <Button onClick={addMockData} >Use mock data</Button>
    </StyledWrapper>
  );
}

export default App;
