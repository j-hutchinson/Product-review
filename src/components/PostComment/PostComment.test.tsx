import Rating from '@material-ui/lab/Rating';
import { shallow } from 'enzyme';
import React, { ChangeEvent } from 'react';
import { PostComment, StyledContainer, StyledInput } from './component';

jest.mock('../../store/store', () => ({ dispatch: jest.fn(), }));

describe('PostComment component', () => {
    const preventDefault = jest.fn();
    const addSingleComment = jest.fn();
    const mockDate = new Date(1466424490000);

    beforeEach(() => {
        jest.spyOn(global, 'Date')
            .mockImplementation(() => mockDate);
    })

    afterEach(() => {
        jest.resetAllMocks();
    });

    test('component matches snapshot', () => {
        expect.assertions(1);

        const wrapper = shallow(<PostComment addSingleComment={addSingleComment} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('onSubmit updates the store with valid Comment object', () => {
        expect.assertions(3);

        const wrapper = shallow(<PostComment addSingleComment={addSingleComment} />);
        const ev = {} as ChangeEvent;

        wrapper.find(StyledInput).at(0).simulate('change', { target: { value: 'Jack Hutchinson' } });
        wrapper.find(StyledInput).at(1).simulate('change', { target: { value: 'email@email.com' } });
        wrapper.find(Rating).props().onChange(ev, 4);
        wrapper.find(StyledInput).at(2).simulate('change', { target: { value: 'This is a good product' } });
        wrapper.find(StyledContainer).props().onSubmit({ preventDefault });

        expect(preventDefault).toHaveBeenCalledTimes(1);
        expect(addSingleComment).toHaveBeenCalledTimes(1);
        expect(addSingleComment).toHaveBeenCalledWith({
            name: 'Jack Hutchinson',
            email: 'email@email.com',
            rating: 4,
            comment: 'This is a good product',
            timePosted: '12:08AM 20/6/2016',
        });
    });
});
