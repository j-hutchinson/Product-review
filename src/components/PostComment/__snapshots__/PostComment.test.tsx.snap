// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostComment component component matches snapshot 1`] = `
<styled.form
  onSubmit={[Function]}
>
  <styled.p>
    Product review for XYZ
  </styled.p>
  <styled.div>
    <label
      htmlFor="name"
    >
      Name: 
    </label>
    <styled.input
      id="name"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </styled.div>
  <styled.div>
    <label
      htmlFor="email"
    >
      Email: 
    </label>
    <styled.input
      id="email"
      onChange={[Function]}
      required={true}
      type="email"
      value=""
    />
  </styled.div>
  <styled.div>
    Rating:
    <styled.div>
      <WithStyles(ForwardRef(Rating))
        name="Product Rating"
        onChange={[Function]}
        value={0}
      />
      <styled.label>
        (
        0
        ) star
        s
      </styled.label>
    </styled.div>
  </styled.div>
  <styled.div>
    <label
      htmlFor="comment"
    >
      Comment
    </label>
    <styled.input
      id="comment"
      onChange={[Function]}
      required={true}
      type="text"
      value=""
    />
  </styled.div>
  <styled.div>
    <styled.input
      type="submit"
      value="Submit"
    />
  </styled.div>
</styled.form>
`;
